spec/formatting.md

```diff
@@ -93,17 +93,32 @@ and different implementations MAY choose to perform different levels of resoluti
 
 Depending on the presence or absence of an _operand_
 and a _function_, _private-use_, or _reserved_ _annotation_,
-one of the following is used to resolve the value of the _expression_:
-
-- If the _expression_ contains no _annotation_,
-  its resolved value is determined by _literal resolution_ or _variable resolution_,
-  depending on the shape of the _operand_.
-- Else, if the _expression_ has a _function_ _annotation_,
-  its resolved value is defined by _function resolution_.
-- Else, if the _expression_ has a _private-use_ _annotation_,
-  its resolved value is defined according to the implementations's specification.
-- Else, the _expression_ has a _reserved_ _annotation_,
-  an Unsupported Expression error is emitted and a fallback value is used as its value.
+the resolved value of the _expression_ is determined as follows:
+
+If the _expression_ contains a _reserved_ _annotation_,
+an `Unsupported Expression` error is emitted and a fallback value is used as its value;
+
+Else, if the _expression_ contains a _private-use_ _annotation_,
+its resolved value is defined according to the implementation's specification;
+
+Else, if the _expression_ contains an _annotation_,
+its resolved value is defined by _function resolution_;
+
+Else, the _expression_ will contain only an _operand_,
+which consists of either a _literal_ or a _variable_.
+
+If the _expression_ consists of a _variable_,
+its resolved value is defined by _variable resolution_.
+An implementation MAY use the value of the _variable_ or
+its resolved value to choose a _function_ and _options_
+to further perform _function resolution_.
```

Comment:
What does "the value of the _variable_ or its resolved value" mean?

Furthermore, this language implicitly restricts an implementation from providing custom behaviour for values e.g. based on their type that is not explicitly made available via _functions_ and _options_. As I argued on the call:
https://github.com/unicode-org/message-format-wg/blob/a53a0619a7dfcb0ce5cec5b1eb99e323d7dbbc85/meetings/2023/notes-2023-08-28.md?plain=1#L345-L349

Reply:
"value or resolved value" is to allow for declaration chaining

```
let $foo = {$someExternalVariableThatIsOfTypeNumber}
let $bar = {$foo}
let $baz = {$bar}
{{$baz} can be a number}
```

How about:

```suggestion
An implementation MAY use the resolved value of a _variable_
to further perform _function resolution_.
```

Note that I used my original permissive language about choosing a function and options because it doesn't make much sense to "apply function resolution" if there is no function! This doesn't forbid an implementation from applying unspecified resolution, such as your person object example. Although, honestly, if you have a `com.foo.person` object and have a "PersonFormatter" function, isn't that "registering the person formatter function"? My text does **not** say "... choose an _annotation_" or "...apply an annotation". It goes directly to the function and its options. I think that leap removes the need for a `:person` keyword. Do we need to say that explicitly?

Reply:
The suggestion is better, but still problematic. The explicit reference to _function resolution_ mean that this step of that process needs to be considered: https://github.com/unicode-org/message-format-wg/blob/92595639fe284c55608b2272b72eeccd8d33a793/spec/formatting.md?plain=1#L147-L151

In other words, this language still requires for something like `:person` to be addressable for any special handling to be assignable to a `com.foo.person` object.

I would prefer something more like this, without an explicit _function resolution_ reference:

```suggestion
An implementation MAY select and apply additional processing
based on the resolved value of a _variable_,
and use the result of such processing as the resolved value of the _expression_.
```

Reply:
What you're suggesting is that if I have a pattern like this:

```
{{$person}}
```

... that an implementation can decide, based on the value of the variable `$person` to supply a formatter function (along with whatever options, i.e. arguments that such a function takes), even though it does not appear in the implementation's "registry" of functions. Note that the registry might not exist as a physical file for a given implementation. 

That seems like a weird feature to have, since it can only be accessed by passing objects that the implementation recognizes and can't be controlled (the pattern can't pass any options). I suppose the formatter could be externally configured, sort of like how ICU4J let's you set a specific formatter in MF1 (generally icky).

In any case, I'm not against allowing some interpolation, but I think gently nudging folks to functions and options in their registry makes sense--if only as a way of making what we're saying understandable. 

How about:

```suggestion
An implementation MAY perform additional processing
when resolving the value of a _variable_ based on the value of the _variable_.
For example, it could apply _function resolution_ using a _function_
and a set of _options_ chosen based on the value or type of the _variable_.
```